File: src/ios/ConsentExecutor.m
#import "ConsentExecutor.h"
#import "AdMobNextGen.h"
#import <CommonCrypto/CommonDigest.h>
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import <AdSupport/AdSupport.h>

@interface ConsentExecutor()
@property (nonatomic, weak) AdMobNextGen *plugin;
@end

@implementation ConsentExecutor

- (instancetype)initWithPlugin:(AdMobNextGen *)plugin {
    self = [super init];
    if (self) {
        self.plugin = plugin;
    }
    return self;
}

- (void)requestConsentInfo:(NSDictionary *)options command:(CDVInvokedUrlCommand *)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        @try {
            BOOL debugMode = options[@"debug"] ? [options[@"debug"] boolValue] : NO;
            NSString *manualTestDeviceId = options[@"testDeviceId"] ? options[@"testDeviceId"] : @"";
            BOOL resetConsent = options[@"reset"] ? [options[@"reset"] boolValue] : NO;
            BOOL tagForUnderAgeOfConsent = options[@"tagForUnderAgeOfConsent"] ? [options[@"tagForUnderAgeOfConsent"] boolValue] : NO;

            if (resetConsent) {
                [UMPConsentInformation.sharedInstance reset];

            }

            UMPRequestParameters *parameters = [[UMPRequestParameters alloc] init];
            parameters.tagForUnderAgeOfConsent = tagForUnderAgeOfConsent;

            if (debugMode) {
                UMPDebugSettings *debugSettings = [[UMPDebugSettings alloc] init];
                debugSettings.geography = UMPDebugGeographyEEA;

                if (manualTestDeviceId.length > 0) {
                    debugSettings.testDeviceIdentifiers = @[manualTestDeviceId];

                } else {
                    NSString *deviceId = [self getDeviceId];
                    if (deviceId) {
                        debugSettings.testDeviceIdentifiers = @[deviceId];

                    }
                }
                parameters.debugSettings = debugSettings;
            }

            [UMPConsentInformation.sharedInstance requestConsentInfoUpdateWithParameters:parameters completionHandler:^(NSError *_Nullable requestConsentError) {

                if (requestConsentError) {

                    [self sendErrorEvent:requestConsentError];
                    if (command) {
                        CDVPluginResult* errRes = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:requestConsentError.localizedDescription];
                        [self.plugin.commandDelegate sendPluginResult:errRes callbackId:command.callbackId];
                    }
                    return;
                }

                [self.plugin fireEvent:@"document" event:@"on.consent.info.update" withData:nil];

                [UMPConsentForm loadAndPresentIfRequiredFromViewController:self.plugin.viewController completionHandler:^(NSError * _Nullable loadAndPresentError) {

                    if (loadAndPresentError) {

                        [self sendErrorEvent:loadAndPresentError];
                        if (command) {
                            CDVPluginResult* errRes = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:loadAndPresentError.localizedDescription];
                            [self.plugin.commandDelegate sendPluginResult:errRes callbackId:command.callbackId];
                        }
                    } else {

                        [self.plugin fireEvent:@"document" event:@"on.consent.form.dismissed" withData:nil];

                        [self sendConsentStatus:command];
                    }
                }];
            }];

        } @catch (NSException *exception) {
            if (command) {
                CDVPluginResult* errRes = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:exception.reason];
                [self.plugin.commandDelegate sendPluginResult:errRes callbackId:command.callbackId];
            }
        }
    });
}

- (void)getTCData:(CDVInvokedUrlCommand *)command {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        @try {
            NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
            NSMutableDictionary *tcData = [NSMutableDictionary dictionary];

            NSString *tcString = [prefs stringForKey:@"IABTCF_TCString"] ?: @"";
            NSString *purposeConsents = [prefs stringForKey:@"IABTCF_PurposeConsents"] ?: @"";
            NSString *purposeLegitimateInterests = [prefs stringForKey:@"IABTCF_PurposeLegitimateInterests"] ?: @""; 
            NSString *vendorConsents = [prefs stringForKey:@"IABTCF_VendorConsents"] ?: @"";
            NSString *additionalConsent = [prefs stringForKey:@"IABTCF_AddtlConsent"] ?: @"";
            NSInteger gdprApplies = [prefs integerForKey:@"IABTCF_gdprApplies"];

            tcData[@"tcString"] = tcString;
            tcData[@"purposeConsents"] = purposeConsents;
            tcData[@"purposeLegitimateInterests"] = purposeLegitimateInterests; 
            tcData[@"vendorConsents"] = vendorConsents;
            tcData[@"additionalConsent"] = additionalConsent;
            tcData[@"gdprApplies"] = @(gdprApplies);

            BOOL isPersonalizedAllowed = NO;
            NSString *statusMessage = @"Unknown";

            BOOL isAdMobPersonalizedAdsAllowed = NO;
            BOOL isAdMobNonPersonalizedAdsAllowed = NO;
            NSString *adMobConsentStatus = @"Unknown";

            if (gdprApplies == 0) {
                isPersonalizedAllowed = YES;
                statusMessage = @"Not GDPR region. Personalized Ads allowed by default.";

                isAdMobPersonalizedAdsAllowed = YES;
                isAdMobNonPersonalizedAdsAllowed = YES;
                adMobConsentStatus = @"Not GDPR region. AdMob ads allowed by default.";
            } else {

                if (purposeConsents.length > 0) {
                    if ([purposeConsents hasPrefix:@"1"]) {
                        isPersonalizedAllowed = YES;
                        statusMessage = @"Purpose 1 Granted. Legacy check passed.";
                    } else {
                        isPersonalizedAllowed = NO;
                        statusMessage = @"Purpose 1 Denied. Legacy check failed.";
                    }
                }

                BOOL hasPurpose1 = [self checkConsent:purposeConsents consentId:1];
                BOOL hasPurpose3 = [self checkConsent:purposeConsents consentId:3];
                BOOL hasPurpose4 = [self checkConsent:purposeConsents consentId:4];

                BOOL hasRequiredLI_or_Consent = 
                    [self hasConsentOrLI:purposeConsents lis:purposeLegitimateInterests consentId:2] &&
                    [self hasConsentOrLI:purposeConsents lis:purposeLegitimateInterests consentId:7] &&
                    [self hasConsentOrLI:purposeConsents lis:purposeLegitimateInterests consentId:9] &&
                    [self hasConsentOrLI:purposeConsents lis:purposeLegitimateInterests consentId:10];

                BOOL hasVendorGoogle = [self checkConsent:vendorConsents consentId:755];

                if (hasPurpose1 && hasVendorGoogle && hasRequiredLI_or_Consent) {
                    isAdMobNonPersonalizedAdsAllowed = YES;

                    if (hasPurpose3 && hasPurpose4) {
                        isAdMobPersonalizedAdsAllowed = YES;
                        adMobConsentStatus = @"Strict requirements met for Personalized Ads (Purposes 1,3,4 + LI 2,7,9,10 + Vendor 755).";
                    } else {
                        isAdMobPersonalizedAdsAllowed = NO;
                        adMobConsentStatus = @"Requirements met for Non-Personalized Ads only.";
                    }
                } else {
                    isAdMobPersonalizedAdsAllowed = NO;
                    isAdMobNonPersonalizedAdsAllowed = NO;
                    adMobConsentStatus = @"Insufficient strict consent (Missing P1, Vendor 755, or P2,7,9,10). Limited Ads only.";
                }
            }

            BOOL isACVersion2 = NO;
            if (additionalConsent.length > 0) {
                isACVersion2 = [additionalConsent hasPrefix:@"2"];
            }

            tcData[@"isPersonalizedAllowed"] = @(isPersonalizedAllowed);
            tcData[@"statusMessage"] = statusMessage;
            tcData[@"isACVersion2"] = @(isACVersion2);

            tcData[@"isAdMobPersonalizedAdsAllowed"] = @(isAdMobPersonalizedAdsAllowed);
            tcData[@"isAdMobNonPersonalizedAdsAllowed"] = @(isAdMobNonPersonalizedAdsAllowed);
            tcData[@"adMobConsentStatus"] = adMobConsentStatus;

            if (command) {
                CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:tcData];
                [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
            }
        } @catch (NSException *exception) {
            if (command) {
                CDVPluginResult* errRes = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:exception.reason];
                [self.plugin.commandDelegate sendPluginResult:errRes callbackId:command.callbackId];
            }
        }
    });
}

- (BOOL)checkConsent:(NSString *)consentString consentId:(NSInteger)consentId {
    if (!consentString || consentString.length < consentId) {
        return NO;
    }
    unichar c = [consentString characterAtIndex:(consentId - 1)];
    return c == '1';
}

- (BOOL)hasConsentOrLI:(NSString *)consents lis:(NSString *)lis consentId:(NSInteger)consentId {
    BOOL hasConsent = [self checkConsent:consents consentId:consentId];
    BOOL hasLI = [self checkConsent:lis consentId:consentId];
    return hasConsent || hasLI;
}

- (void)showPrivacyOptionsForm:(CDVInvokedUrlCommand *)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        [UMPConsentForm presentPrivacyOptionsFormFromViewController:self.plugin.viewController completionHandler:^(NSError * _Nullable formError) {
            if (formError) {
                [self sendErrorEvent:formError];
                if (command) {
                    CDVPluginResult* errRes = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:formError.localizedDescription];
                    [self.plugin.commandDelegate sendPluginResult:errRes callbackId:command.callbackId];
                }
            } else {
                [self.plugin fireEvent:@"document" event:@"on.consent.form.dismissed" withData:nil];
                [self sendConsentStatus:command];
            }
        }];
    });
}

- (void)canRequestAds:(CDVInvokedUrlCommand *)command {
    BOOL canRequest = UMPConsentInformation.sharedInstance.canRequestAds;
    CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:canRequest ? 1 : 0];
    [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
}

#pragma mark - Helpers

- (void)sendConsentStatus:(CDVInvokedUrlCommand *)command {
    NSMutableDictionary *result = [NSMutableDictionary dictionary];

    BOOL canRequestAds = UMPConsentInformation.sharedInstance.canRequestAds;
    result[@"canRequestAds"] = @(canRequestAds);

    UMPPrivacyOptionsRequirementStatus reqStatus = UMPConsentInformation.sharedInstance.privacyOptionsRequirementStatus;

    NSString *reqStatusString = @"UNKNOWN";
    if (reqStatus == UMPPrivacyOptionsRequirementStatusRequired) reqStatusString = @"REQUIRED";
    else if (reqStatus == UMPPrivacyOptionsRequirementStatusNotRequired) reqStatusString = @"NOT_REQUIRED";

    result[@"privacyOptionsRequirementStatus"] = reqStatusString;
    result[@"isPrivacyOptionsRequired"] = @(reqStatus == UMPPrivacyOptionsRequirementStatusRequired);
    result[@"consentStatus"] = @(UMPConsentInformation.sharedInstance.consentStatus);

    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:result options:0 error:&error];
    if (!error && jsonData) {
        NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        [self.plugin fireEvent:@"document" event:@"on.consent.status.change" withData:jsonString];
    }

    if (command) {
        CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];
        [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
    }
}

- (void)sendErrorEvent:(NSError *)error {
    NSString *jsonStr = [NSString stringWithFormat:@"{\"code\":%ld, \"message\":\"%@\"}", (long)error.code, error.localizedDescription];
    [self.plugin fireEvent:@"document" event:@"on.consent.error" withData:jsonStr];
}

- (NSString *)getDeviceId {

    NSString *uuid = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    if (!uuid) return nil;

    const char *cStr = [uuid UTF8String];
    unsigned char digest[CC_SHA256_DIGEST_LENGTH]; 
    CC_SHA256(cStr, (CC_LONG)strlen(cStr), digest);

    NSMutableString *output = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for(int i = 0; i < CC_SHA256_DIGEST_LENGTH; i++)
        [output appendFormat:@"%02X", digest[i]];

    return [output uppercaseString];
}

#pragma mark - App Tracking Transparency (ATT / IDFA)

- (void)requestTrackingAuthorization:(CDVInvokedUrlCommand *)command {
    if (@available(iOS 14.0, *)) {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {

            NSString *statusString = [self getATTStatusString:status];

            NSString *jsonStr = [NSString stringWithFormat:@"{\"status\":\"%@\"}", statusString];
            [self.plugin fireEvent:@"document" event:@"on.consent.att.status" withData:jsonStr];

            if (command) {
                CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:statusString];
                [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
            }
        }];
    } else {

        NSString *statusString = @"AUTHORIZED";
        if (command) {
            CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:statusString];
            [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        }
    }
}

- (void)getTrackingAuthorizationStatus:(CDVInvokedUrlCommand *)command {
    if (@available(iOS 14.0, *)) {
        ATTrackingManagerAuthorizationStatus status = [ATTrackingManager trackingAuthorizationStatus];
        NSString *statusString = [self getATTStatusString:status];

        if (command) {
            CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:statusString];
            [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        }
    } else {
        if (command) {
            CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"AUTHORIZED"];
            [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        }
    }
}

- (NSString *)getATTStatusString:(NSUInteger)status {
    if (@available(iOS 14.0, *)) {
        switch (status) {
            case ATTrackingManagerAuthorizationStatusAuthorized:
                return @"AUTHORIZED";
            case ATTrackingManagerAuthorizationStatusDenied:
                return @"DENIED";
            case ATTrackingManagerAuthorizationStatusNotDetermined:
                return @"NOT_DETERMINED";
            case ATTrackingManagerAuthorizationStatusRestricted:
                return @"RESTRICTED";
            default:
                return @"UNKNOWN";
        }
    }
    return @"AUTHORIZED";
}

@end

File: src/android/ConsentExecutor.java
package com.emi.cordova.admob.nextgen;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.util.Log;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView; 
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Locale;

import com.google.android.ump.ConsentDebugSettings;
import com.google.android.ump.ConsentInformation;
import com.google.android.ump.ConsentRequestParameters;
import com.google.android.ump.FormError;
import com.google.android.ump.UserMessagingPlatform;

public class ConsentExecutor {

    private static final String TAG = "AdMobConsent";
    private CordovaInterface cordova;
    private CordovaWebView webView; 
    private ConsentInformation consentInformation;

    public ConsentExecutor(CordovaInterface cordova, CordovaWebView webView) {
        this.cordova = cordova;
        this.webView = webView;
        this.consentInformation = UserMessagingPlatform.getConsentInformation(cordova.getActivity());
    }

    public void requestConsentInfo(JSONArray args, CallbackContext callbackContext) {
            try {
                JSONObject options = args.optJSONObject(0);

                boolean debugMode = false;
                String manualTestDeviceId = "";
                boolean resetConsent = false;
                boolean tagForUnderAgeOfConsent = false;

                if (options != null) {
                    debugMode = options.optBoolean("debug", false);
                    manualTestDeviceId = options.optString("testDeviceId", "");
                    resetConsent = options.optBoolean("reset", false);

                    if (options.has("tagForUnderAgeOfConsent")) {
                        tagForUnderAgeOfConsent = options.getBoolean("tagForUnderAgeOfConsent");
                    }
                }

                if (resetConsent) {
                    consentInformation.reset();

                }

                ConsentRequestParameters.Builder paramsBuilder = new ConsentRequestParameters.Builder();
                paramsBuilder.setTagForUnderAgeOfConsent(tagForUnderAgeOfConsent);

                if (debugMode) {
                    ConsentDebugSettings.Builder debugSettingsBuilder = new ConsentDebugSettings.Builder(cordova.getActivity())
                            .setDebugGeography(ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA);

                    if (!manualTestDeviceId.isEmpty()) {
                        debugSettingsBuilder.addTestDeviceHashedId(manualTestDeviceId);

                    } else {
                        String deviceId = getDeviceId();
                        if (deviceId != null) {
                            debugSettingsBuilder.addTestDeviceHashedId(deviceId);

                        }
                    }

                    paramsBuilder.setConsentDebugSettings(debugSettingsBuilder.build());
                }

                ConsentRequestParameters params = paramsBuilder.build();
                Activity activity = cordova.getActivity();

                activity.runOnUiThread(() -> {
                    consentInformation.requestConsentInfoUpdate(
                            activity,
                            params,
                            () -> {

                                fireEvent("on.consent.info.update", null);

                                UserMessagingPlatform.loadAndShowConsentFormIfRequired(
                                        activity,
                                        (FormError loadAndShowError) -> {
                                            if (loadAndShowError != null) {

                                                sendErrorEvent(loadAndShowError);
                                                callbackContext.error(loadAndShowError.getMessage());
                                            } else {

                                                fireEvent("on.consent.form.dismissed", null);

                                                sendConsentStatus(callbackContext);
                                            }
                                        }
                                );
                            },
                            (FormError requestConsentError) -> {

                                sendErrorEvent(requestConsentError);
                                callbackContext.error(requestConsentError.getMessage());
                            }
                    );
                });

            } catch (Exception e) {
                callbackContext.error("Exception: " + e.getMessage());
            }
    }

    public void getTCData(CallbackContext callbackContext) {
        try {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(cordova.getActivity());
            JSONObject tcData = new JSONObject();

            String tcString = prefs.getString("IABTCF_TCString", "");
            String purposeConsents = prefs.getString("IABTCF_PurposeConsents", "");
            String purposeLegitimateInterests = prefs.getString("IABTCF_PurposeLegitimateInterests", ""); 
            String vendorConsents = prefs.getString("IABTCF_VendorConsents", "");
            int gdprApplies = prefs.getInt("IABTCF_gdprApplies", 0);

            tcData.put("tcString", tcString);
            tcData.put("purposeConsents", purposeConsents);
            tcData.put("purposeLegitimateInterests", purposeLegitimateInterests); 
            tcData.put("vendorConsents", vendorConsents);
            tcData.put("gdprApplies", gdprApplies);

            boolean isPersonalizedAllowed = false;
            String statusMessage = "Unknown";

            boolean isAdMobPersonalizedAdsAllowed = false;
            boolean isAdMobNonPersonalizedAdsAllowed = false;
            String adMobConsentStatus = "Unknown";

            if (gdprApplies == 0) {

                isPersonalizedAllowed = true;
                statusMessage = "Not GDPR region. Personalized Ads allowed by default.";

                isAdMobPersonalizedAdsAllowed = true;
                isAdMobNonPersonalizedAdsAllowed = true;
                adMobConsentStatus = "Not GDPR region. AdMob ads allowed by default.";
            } else {

                if (purposeConsents != null && purposeConsents.length() > 0) {
                    char p1 = purposeConsents.charAt(0);
                    if (p1 == '1') {
                        isPersonalizedAllowed = true;
                        statusMessage = "Purpose 1 Granted. Legacy check passed.";
                    } else {
                        isPersonalizedAllowed = false;
                        statusMessage = "Purpose 1 Denied. Legacy check failed.";
                    }
                }

                boolean hasPurpose1 = checkConsent(purposeConsents, 1);
                boolean hasPurpose3 = checkConsent(purposeConsents, 3);
                boolean hasPurpose4 = checkConsent(purposeConsents, 4);

                boolean hasRequiredLI_or_Consent = 
                    hasConsentOrLI(purposeConsents, purposeLegitimateInterests, 2) &&
                    hasConsentOrLI(purposeConsents, purposeLegitimateInterests, 7) &&
                    hasConsentOrLI(purposeConsents, purposeLegitimateInterests, 9) &&
                    hasConsentOrLI(purposeConsents, purposeLegitimateInterests, 10);

                boolean hasVendorGoogle = checkConsent(vendorConsents, 755); 

                if (hasPurpose1 && hasVendorGoogle && hasRequiredLI_or_Consent) {
                    isAdMobNonPersonalizedAdsAllowed = true;

                    if (hasPurpose3 && hasPurpose4) {
                        isAdMobPersonalizedAdsAllowed = true;
                        adMobConsentStatus = "Strict requirements met for Personalized Ads (Purposes 1,3,4 + LI 2,7,9,10 + Vendor 755).";
                    } else {
                        isAdMobPersonalizedAdsAllowed = false;
                        adMobConsentStatus = "Requirements met for Non-Personalized Ads only.";
                    }
                } else {
                    isAdMobPersonalizedAdsAllowed = false;
                    isAdMobNonPersonalizedAdsAllowed = false;
                    adMobConsentStatus = "Insufficient strict consent (Missing P1, Vendor 755, or P2,7,9,10). Limited Ads only.";
                }
            }

            tcData.put("isPersonalizedAllowed", isPersonalizedAllowed);
            tcData.put("statusMessage", statusMessage);

            tcData.put("isAdMobPersonalizedAdsAllowed", isAdMobPersonalizedAdsAllowed);
            tcData.put("isAdMobNonPersonalizedAdsAllowed", isAdMobNonPersonalizedAdsAllowed);
            tcData.put("adMobConsentStatus", adMobConsentStatus);

            callbackContext.success(tcData);
        } catch (Exception e) {
            callbackContext.error("Failed to read TC Data: " + e.getMessage());
        }
    }

    private boolean checkConsent(String consentString, int id) {
        if (consentString == null || consentString.length() < id) {
            return false;
        }
        return consentString.charAt(id - 1) == '1';
    }

    private boolean hasConsentOrLI(String consents, String lis, int id) {
        boolean hasConsent = checkConsent(consents, id);
        boolean hasLI = checkConsent(lis, id);
        return hasConsent || hasLI;
    }

    private String getDeviceId() {
        try {
            Context context = cordova.getActivity();
            @SuppressLint("HardwareIds") String androidId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
            if (androidId == null || androidId.isEmpty()) return null;
            return md5(androidId).toUpperCase(Locale.getDefault());
        } catch (Exception e) {
            return null;
        }
    }

    private String md5(String s) {
        try {
            MessageDigest digest = MessageDigest.getInstance("MD5");
            digest.update(s.getBytes());
            byte messageDigest[] = digest.digest();
            StringBuilder hexString = new StringBuilder();
            for (byte b : messageDigest) {
                String h = Integer.toHexString(0xFF & b);
                while (h.length() < 2) h = "0" + h;
                hexString.append(h);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            return "";
        }
    }

    private void sendConsentStatus(CallbackContext callbackContext) {
        try {
            JSONObject result = new JSONObject();
            boolean canRequestAds = consentInformation.canRequestAds();
            result.put("canRequestAds", canRequestAds);

            ConsentInformation.PrivacyOptionsRequirementStatus requirementStatus =
                    consentInformation.getPrivacyOptionsRequirementStatus();
            result.put("privacyOptionsRequirementStatus", requirementStatus.name());
            result.put("isPrivacyOptionsRequired",
                    requirementStatus == ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED);

            result.put("consentStatus", consentInformation.getConsentStatus());

            fireEvent("on.consent.status.change", result);

            if (callbackContext != null) {
                callbackContext.success(result);
            }
        } catch (JSONException e) {
            if (callbackContext != null) callbackContext.error("JSON Error");
        }
    }

    public void canRequestAds(CallbackContext callbackContext) {
        callbackContext.success(consentInformation.canRequestAds() ? 1 : 0);
    }

    public void showPrivacyOptionsForm(CallbackContext callbackContext) {
        Activity activity = cordova.getActivity();
        cordova.getActivity().runOnUiThread(() -> {
            UserMessagingPlatform.showPrivacyOptionsForm(
                    activity,
                    (FormError formError) -> {
                        if (formError != null) {
                            sendErrorEvent(formError);
                            callbackContext.error(formError.getMessage());
                        } else {

                            fireEvent("on.consent.form.dismissed", null);
                            sendConsentStatus(callbackContext);
                        }
                    }
            );
        });
    }

    private void sendErrorEvent(FormError error) {
        try {
            JSONObject errData = new JSONObject();
            errData.put("code", error.getErrorCode());
            errData.put("message", error.getMessage());
            fireEvent("on.consent.error", errData);
        } catch (JSONException e) {}
    }

    private void fireEvent(String eventName, JSONObject data) {
        cordova.getActivity().runOnUiThread(() -> {
            StringBuilder js = new StringBuilder();
            js.append("javascript:cordova.fireDocumentEvent('");
            js.append(eventName);
            js.append("'");

            if (data != null) {
                js.append(", ");
                js.append(data.toString());
            }

            js.append(");");

            String jsCommand = js.toString();
            if (webView != null) webView.loadUrl(jsCommand);
        });
    }
}
